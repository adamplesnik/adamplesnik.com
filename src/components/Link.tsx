import { motion } from 'framer-motion'
import { ArrowRight, LucideIcon } from 'lucide-react'
import { PropsWithChildren, useState } from 'react'

const Link = ({ Icon = undefined, href, children }: PropsWithChildren<LinkProps>) => {
  const [anim, setAnim] = useState(false)

  return (
    <div
      className="relative mb-2 flex gap-2"
      onMouseEnter={() => setAnim(true)}
      onMouseLeave={() => setAnim(false)}
    >
      {Icon ? (
        <motion.div
          animate={{
            scaleY: anim ? 0 : 1,
            y: anim ? 12 : 0,
            // opacity: anim ? 0 : 1,
            width: anim ? 0 : undefined,
          }}
          transition={{ duration: 0.3, delay: anim ? 0 : 0.45 }}
        >
          <Icon />
        </motion.div>
      ) : (
        ''
      )}
      <motion.div layout className="truncate">
        <a href={href} target="_blank" className="truncate">
          {children}
        </a>
      </motion.div>
      <motion.div
        animate={{
          rotate: anim ? 0 : -45,
          scale: anim ? 1 : 0,
          y: anim ? 0 : 12,
          width: anim ? undefined : 0,
        }}
        transition={{ delay: 0.48 }}
      >
        <ArrowRight />
      </motion.div>
      <motion.div
        animate={{ left: anim ? '88%' : 12, width: anim ? 128 : 0 }}
        transition={{ delay: 0.27, duration: 0.56, ease: [0, 1, 0, 1] }}
        className={
          'absolute -bottom-1 left-0 h-0.5 bg-gradient-to-r from-fuchsia-400 to-violet-800 transition-[max-width] duration-500 '
        }
      ></motion.div>
    </div>
  )
}

export interface LinkProps {
  href: string
  children: PropsWithChildren
  Icon?: LucideIcon | undefined
}

export default Link
