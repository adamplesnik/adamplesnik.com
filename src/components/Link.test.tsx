import { toJson } from '@/utils/toJson'
import renderer from 'react-test-renderer'
import { expect, test } from 'vitest'
import Link from './Link'

test('Link with href', () => {
  const component = renderer.create(<Link href="url/">link</Link>)
  const tree = toJson(component)
  expect(tree).toMatchSnapshot()
})

test('Link with href and custom target', () => {
  const component = renderer.create(
    <Link href="url/" target="_self">
      link
    </Link>
  )
  const tree = toJson(component)
  expect(tree).toMatchSnapshot()
})

test('Link with href, custom target and icon', () => {
  const component = renderer.create(
    <Link href="url/" target="_self">
      link
    </Link>
  )
  const tree = toJson(component)
  expect(tree).toMatchSnapshot()
})
