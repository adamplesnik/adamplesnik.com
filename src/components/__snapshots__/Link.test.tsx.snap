// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Link with href 1`] = `
<div
  className="relative flex items-center"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <a
    className="bg-gradient-to-r from-fuchsia-500 to-indigo-400 bg-clip-text text-transparent hover:to-fuchsia-700 print:bg-transparent print:text-gray-700"
    href="url/"
    target="_blank"
  >
    link
  </a>
</div>
`;

exports[`Link with href and custom target 1`] = `
<div
  className="relative flex items-center"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <a
    className="bg-gradient-to-r from-fuchsia-500 to-indigo-400 bg-clip-text text-transparent hover:to-fuchsia-700 print:bg-transparent print:text-gray-700"
    href="url/"
    target="_self"
  >
    link
  </a>
</div>
`;

exports[`Link with href, custom target and icon 1`] = `
<div
  className="relative flex items-center"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    style={
      {
        "transform": "none",
        "willChange": "transform",
      }
    }
  >
    <svg
      className="lucide lucide-cog size-7 pr-2 text-fuchsia-600 opacity-70"
      fill="none"
      height={24}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={1.8}
      viewBox="0 0 24 24"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 20a8 8 0 1 0 0-16 8 8 0 0 0 0 16Z"
      />
      <path
        d="M12 14a2 2 0 1 0 0-4 2 2 0 0 0 0 4Z"
      />
      <path
        d="M12 2v2"
      />
      <path
        d="M12 22v-2"
      />
      <path
        d="m17 20.66-1-1.73"
      />
      <path
        d="M11 10.27 7 3.34"
      />
      <path
        d="m20.66 17-1.73-1"
      />
      <path
        d="m3.34 7 1.73 1"
      />
      <path
        d="M14 12h8"
      />
      <path
        d="M2 12h2"
      />
      <path
        d="m20.66 7-1.73 1"
      />
      <path
        d="m3.34 17 1.73-1"
      />
      <path
        d="m17 3.34-1 1.73"
      />
      <path
        d="m11 13.73-4 6.93"
      />
    </svg>
  </div>
  <a
    className="bg-gradient-to-r from-fuchsia-500 to-indigo-400 bg-clip-text text-transparent hover:to-fuchsia-700 print:bg-transparent print:text-gray-700"
    href="url/"
    target="_self"
  >
    link
  </a>
</div>
`;
